### ISSUE TITLE
Question:Can I sync data between IndexedDB & SQL database?

### ISSUE THREAD

**Issue by wktdev** [dfahlander/Dexie.js#1168]
I am building an app that lets users input data and store that data offline in IndexedDB via Dexie.js.
When the users authenticates, I want the data that they've entered while unauthenticated to be merged with any data they have stored in their user account. If they happen to go offline while working, when they get back online I want the data to resolveou

I've read that tools like puchDB with CouchDB are intended for this kind of use case.

I want to know if Dexie has any built in tools that will allow me to do the same thing but with IndexedDB & an SQL database.

Thank you

**Comment by dfahlander** [comment]
Have a look at [Dexie Cloud](https://dexie.org/cloud/) as it will enable this exact scenario. It's not free though (but hosting is normally not free). The downside with CouchDB is that it doesn't have any separation of users, so you may end up with one db per user. Same restriction applies to [sync-client](https://www.npmjs.com/package/sync-client) and [sync-server](https://www.npmjs.com/package/sync-server) which together builds a Dexie-based sync solution.

You can also implement this yourself using [Dexie.Syncable.js](https://dexie.org/docs/Syncable/Dexie.Syncable.js) or possibly write your own Dexie hooks and middlewares the same way Dexie.Syncable does. I've used Dexie.Syncable successfully in an old app that I'm still using, where I'm using a .NET backend with SQLServer. Dexie.Syncable is in beta and has been so for a very long time. However, I will be working on a rewrite and a stable version before [Dexie Cloud](https://dexie.org/cloud/) is released in stable version.

**Comment by wktdev** [comment]
Thank you. The instructions at [Dexie.Syncavle.js](https://dexie.org/docs/Syncable/Dexie.Syncable.js) look very complicated. Is there any tangible working examples that I can download and get working with a simple npm install ? 

So alternatively I looked at Dexie Cloud and I read this line: 
**With this very simple and default setup, users will be requested to login using passwordless email OTP the very first time they visit your app**

I don't want users to have to sign up when they visit my app. If they want to  persist their data, then they will sign up as a paid monthly subscriber. It's an important part of the UI to allow users to use my app without any sign up whatsoever. 

Is what I am describing possible using Dexie Cloud?
Thanks again

**Comment by dfahlander** [comment]
Yes, that's a scenario in dexie cloud by setting [{ requireAuth: false }](https://dexie.org/cloud/docs/db.cloud.configure()#requireauth) in the options to db.cloud.configure().

The scenario is this:

1. User loads your app. 
2. Cloud is connected to anonymously and if there is any data in the [public realm](https://dexie.org/cloud/docs/access-control#the-public-realm) this will be synced down to the client.
3. User interacts with your application and can create and mutate own data but that data is not yet synced to server.
4. User actively authenticates. Now the user data can be either synced up to cloud or thrown away and replaced by the persisted data for that user. This depends on the use case and should be possible for your application or user to decide. The API for this part is not yet set.

**Comment by dheimoz** [comment]
Hello @dfahlander I must admit that Dexie and Dexie Cloud are great technologies. The principle it appealing. However, I believe your comments regarding the downsides about Couchdb are misleading.

https://medium.com/@eiri/couchdb-authorization-in-a-database-58c8ee633c96

Cheers.

**Comment by dfahlander** [comment]
Thanks @dheimoz for pointing out the comment that may be misleading. Couch has separation of users indeed, but what is unique to dexie cloud is the realm- (and document- /row-) level access control and that the application developer can base it on an authentication mechanism of choice.

**Comment by dheimoz** [comment]
All right @dfahlander . Your work is definitely worth it. I am active supporter of Pouchdb and Couchdb but I am ready to use your technology.

I wanted to ask about attachments, would you want me to open a new issue and if @wktdev is ok maybe close this issue?

Thanks.

**Comment by alexander-potemkin** [comment]
I know it's been a while, but I have the same question :)

Is it possible to sync Dexie with PostgreSQL database? The thing is that we already have it in the project and I would like to avoid extra moving parts, whenever possible.

**Comment by frasmarco** [comment]
@alexander-potemkin same requirement for my side, did you found a solution?

**Comment by dfahlander** [comment]
> Is it possible to sync Dexie with PostgreSQL database? The thing is that we already have it in the project and I would like to avoid extra moving parts, whenever possible.

Dexie Cloud stores its data in a standard Postgres database in the backend. The [cloud service](https://dexie.org/cloud/pricing#prod-edition) is not exposing the postgres database though so it needs to be queried over [its REST API](https://dexie.org/cloud/docs/rest-api), [the CLI](https://dexie.org/cloud/docs/cli) or the sync endpoint that is used by [dexie-cloud-addon](https://dexie.org/cloud/docs/dexie-cloud-addon). Customers of the [Silver](https://dexie.org/cloud/pricing#silver-edition) and [Gold](https://dexie.org/cloud/pricing#gold-edition) editions connect the server to their own postgres database though and can query it directly.

> I wanted to ask about attachments, ...

Regarding attachements, dexie cloud will store Blob properties in an alternate blob storage instead of Postgres in an upcoming release some time in summer. There won't be any action needed from the app developer's side since it will just optimize how blobs are being stored and synced and that happens in the background anyway, so I advice to save large binaries in blobs now to allow for optimized blob handling once it's released.

So basically, if you have attachments, you could store them as Blobs. For example, have a dexie table where the attachements are stored:

```ts
const db = new Dexie('mydb', { addons: [ dexieCloudAddon ] });
db.version(1).stores({
  ...,
  attachements: '@id'
});

async function storeAttachment(blob: Blob) {
  const id = await db.attachements.add({blob});
  return id;
}

```

**Comment by Alamin-Juma** [comment]
> I am building an app that lets users input data and store that data offline in IndexedDB via Dexie.js. When the users authenticates, I want the data that they've entered while unauthenticated to be merged with any data they have stored in their user account. If they happen to go offline while working, when they get back online I want the data to resolve
> 
> I've read that tools like puchDB with CouchDB are intended for this kind of use case.
> 
> I want to know if Dexie has any built in tools that will allow me to do the same thing but with IndexedDB & an SQL database. If so, could you point me in the right direction.
> 
> Thank you

Guys don't worry, am your angel here. Have a look at this video. Thank me and please buy me some coffee by sending me an email to alaminjuma96@gmail.com. You welcome https://www.youtube.com/watch?v=lrzRGyBeWpQ&ab_channel=AldoCaamal
